monitor/contract/monitor.go

```diff
@@ -0,0 +1,93 @@
+package contract
+
+import (
+	"context"
+	"time"
+
+	"github.com/omni-network/omni/lib/contracts"
+	"github.com/omni-network/omni/lib/ethclient"
+	"github.com/omni-network/omni/lib/log"
+	"github.com/omni-network/omni/lib/netconf"
+
+	"github.com/ethereum/go-ethereum/params"
+)
+
+// StartMonitoring starts the monitoring goroutines.
+func StartMonitoring(ctx context.Context, network netconf.Network, rpcClients map[uint64]ethclient.Client) {
+	log.Info(ctx, "Monitoring contracts")
+
+	for _, chain := range network.EVMChains() {
+		isOmniEVM := chain.ID == network.ID.Static().OmniExecutionChainID
+
+		for _, contract := range contracts.ToFund(network.ID) {
+			if contract.OnlyOmniEVM && !isOmniEVM {
+				continue
+			}
+
+			go monitorContractForever(ctx, contract, chain.Name, rpcClients[chain.ID])
+		}
+	}
+}
+
+// monitorContractForever blocks and periodically monitors the contract for the given chain.
+func monitorContractForever(
+	ctx context.Context,
+	contract contracts.WithFundThreshold,
+	chainName string,
+	client ethclient.Client,
+) {
+	ctx = log.WithCtx(ctx,
+		"chain", chainName,
+		"name", contract.Name,
+		"address", contract.Address,
+	)
+
+	log.Info(ctx, "Monitoring account")
+
+	ticker := time.NewTicker(time.Second * 30)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			err := monitorContractOnce(ctx, contract, chainName, client)
+			if ctx.Err() != nil {
+				return
+			} else if err != nil {
+				log.Error(ctx, "Monitoring contract failed (will retry)", err)
```

Comment:
We try to use WARN for stuff that retries, and only use ERROR for stuff that can't retry and require operator intervention